### PR TITLE
supoport canary deployment

### PR DESCRIPTION
**Pull request by cvvz**
### What problem does this PR solve? 
https://github.com/pingcap/tidb-operator/pull/3243
https://github.com/pingcap/tidb-operator/pull/3166
https://github.com/pingcap/tidb-operator/pull/1866

### What is changed and how does it work?
This PR is merged with https://github.com/pingcap/tidb-operator/pull/3548 ,  @shonge and I did this together. **Please review the code of `helm` part in his PR https://github.com/pingcap/tidb-operator/pull/3548.**
Users now can deploy mutiple operators(tidb-controller-manager and tidb-scheduler) in k8s by helm. And tidb-controller-manager will work independently by filtering the CR by label.

### Check List 

Tests 

- Manual test 
  - Install two operator releases by helm with values of `scheduler.serviceAccount`, `scheduler.schedulerName`, `controllerManager.serviceAccount` different and only install `controllerManager` and `scheduler`. (`advancedStatefulset` and `admissionWebhook` is not allowed to install mutiple) And one controller is set the `controllerManager.selector` as `k1=v1`, another is set as `k1!=v1`.
  - Create two `TidbCluster`, one is labeled with `k1=v1` and set the scheduler name as `tidb-scheduler-1`, another is not labeled and set the scheduler name as `tidb-scheduler-2`.
  - Notice that each `TidbCluster` will **only** be watched by specific controller.

Code changes

- Has Go code change

Related changes

- Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support configuring `--selector` for the tidb-controller-manager to manage specific CRs 
```

### CODE REVIEW

**Comment by dragonly** [comment]
Could you please complete the PR template and make a proper description of all the required answers there?

Maybe take #3517 as an example 😉

**Comment by cvvz** [comment]
> Could you please complete the PR template and make a proper description of all the required answers there?
> 
> Maybe take #3517 as an example 😉

@dragonly PTAL 😄

**Comment by cvvz** [comment]
cc @DanielZhangQD

File: cmd/controller-manager/main.go [hunk 1]
```diff
@@ -186,7 +191,7 @@ func main() {
 			Lock: &resourcelock.EndpointsLock{
 				EndpointsMeta: metav1.ObjectMeta{
 					Namespace: ns,
-					Name:      "tidb-controller-manager",
+					Name:      helmRelease + "-controller-manager",
```
Comment (DanielZhangQD) [inline cmd/controller-manager/main.go:194]:
```suggestion
					Name:      "tidb-controller-manager-"helmRelease,
```

**Comment by DanielZhangQD** [comment]
@cvvz Please remove the code change for charts from your PR.

**Comment by DanielZhangQD** [comment]
@cvvz Please also fix the ci.
BTW, I created an e2e issue https://github.com/pingcap/tidb-operator/issues/3566 for this feature, could you please add e2e case in that issue?

**Comment by cvvz** [comment]
> BTW, I created an e2e issue #3566 for this feature, could you please add e2e case in that issue?

@DanielZhangQD The e2e test will be done after I finish the work in company, maybe next month.

**Comment by dragonly** [comment]
@cvvz  I think this PR introduces an impressive user-facing feature, so could you please also update the **release note** in the PR description?
You can also reference to the [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md)

**Comment by cvvz** [comment]
@dragonly DONE. But I can't add `release-note-action-required` label to this PR.

**Comment by DanielZhangQD** [comment]
@cvvz Please also help add doc for this feature in the docs-tidb-operator repo, thanks!

**Comment by DanielZhangQD** [comment]
@cvvz Did you test the following CRs and confirm that they will not be managed by two TiDB Operators?

- TidbCluster 
- TidbMonitor 
- TidbInitializer 
- Backup 
- Restore 
- BackupSchedule 
- TidbClusterAutoScaler 
- DMCluster

Could you please update the test result in the PR description?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by mikechengwei** [comment]
plz update branch

File: cmd/controller-manager/main.go [hunk 2]
```diff
@@ -186,7 +191,7 @@ func main() {
 			Lock: &resourcelock.EndpointsLock{
 				EndpointsMeta: metav1.ObjectMeta{
 					Namespace: ns,
-					Name:      "tidb-controller-manager",
+					Name:      "tidb-controller-manager-" + helmRelease,
```
Comment (mikechengwei) [inline cmd/controller-manager/main.go:194]: There are compatibility issues here when upgrade operator.

File: cmd/controller-manager/main.go [hunk 3]
```diff
@@ -79,6 +79,11 @@ func main() {
 		klog.Fatal("NAMESPACE environment variable not set")
 	}
 
+	helmRelease := os.Getenv("HELM_RELEASE")
+	if helmRelease == "" {
+		klog.Fatal("HELM_RELEASE environment variable not set")
```
Comment (DanielZhangQD) [inline cmd/controller-manager/main.go:84]: Is it OK that if this env is not set, we just use the original endpoint name in L194?

File: cmd/controller-manager/main.go (hunk 2 above)
Comment (cvvz) [reply cmd/controller-manager/main.go]: @mikechengwei If `helmRelease` is set, a new endpoint with suffix will be created, and the old one won't be deleted, both of them can work normally. In which situation do you think will cause compatibility issues? I don't get it , please explain with more detail, thanks a lot 😄

Comment (mikechengwei) [reply cmd/controller-manager/main.go]: There is nothing wrong with this piece of code, just set the default value, compatible with the old logic, the new commit should be no problem.

Comment (cvvz) [reply cmd/controller-manager/main.go]: Got it

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by mikechengwei** [comment]
LGTM

**Review by mikechengwei — APPROVED** [review]

**Comment by mikechengwei** [comment]
LGTM

**Pull request merged**